src/main/java/com/flab/bbt/product/repository/ProductRepositoryImpl.java

```diff
@@ -0,0 +1,36 @@
+package com.flab.bbt.product.repository;
+
+import com.flab.bbt.product.domain.Product;
+import com.flab.bbt.user.domain.User;
+import org.springframework.stereotype.Repository;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+@Repository
+public class ProductRepositoryImpl implements ProductRepository{
+    private static Map<Long, Product> productDb = new HashMap<>();
+    private static Map<String, Long> serialNoIndex = new HashMap<>();
+    private static long sequence = 0L;
+
+    @Override
+    public Product save(Product product) {
+        product.setId(++sequence);
+        productDb.put(product.getId(), product);
+        serialNoIndex.put(product.getSerialNum(), product.getId());
+
```

Comment:
동시성 문제가 발생할수 있을것 같습니다.
thread 2개가 있다고 가정해봅시다.(T1,T2)
T1, T2에서 같은 `sequence`를 발급받아 `productDb.put(product.getId(), product);` 가 실행된다면
나중에 실행된 데이터만 보존되고 이전 데이터는 덮어씌워질 수 있습니다.

이를 예방하려면 어떡헤 해야할까요?

Reply:
productDb를 ConcurrentHashMap 로 관리해야 할 것 같습이다! 수정하겠습니다ㅎㅎ

Reply:
@jungin-jin-choi 
`ConcurrentHashMap`을 사용한다 하더라도 제가 언급드린 이슈는 계속 발생할것 같습니다.
해결 할 수 있는 다른방법이 있을까요?

Reply:
오 제가 동시성에 대한 이해가 부족해서말씀하신 상황을 첨에 제대로 못 이해했었네요.
1. ConcurrentHashMap으로 바꿀 경우, producctDb.put을 할 때 락을 걸기 때문에, WRITE 자체는 동시에 한 스레드만이 할 수 있다. (ConcurrentHashMap의 put 자체는 thread safe하다)
2. 하지만 서로 다른 스레드가 같은 sequence를 발급받은 경우, productDb.put가 차례로 진행되면 데이터가 덮어씌워질 수 있다. (하지만 call 자체는 threadsafe하지 않을 수 있다)

해결하려면.. 
아래와 같이 할 수 있을 것 같습니다.
```suggestion
product.setId(++sequence);
if(productDb.putIfAbsent(product.getId(), product)==null){
   serialNoIndex.put(p.getSerialNum(), p.getId());
}
```
product.setId(++sequence);
productDb.put(product.getId(), product);
serialNoIndex.put(product.getSerialNum(), product.getId());

이렇게 할 경우 p에 결국 하나의 Product 객체가 들어가게 되어서 productDb와 serialNoIndex가 잘못 업데이트되는 경우를 방지할 수 있을 것 같습니다..!

tmi지만 저 오늘 회사에서도 락을 처음으로 걸어봤는데 오늘 동시성의 날인가봅니다,, 동시성 약한데 잘됐네욤,,

Reply:
@jungin-jin-choi 
여전히 해결이 되지 않았습니다!
덕분에 동시성으로 할 수 있는 많은 공부를 하게되었군요😀

위와같이 처리하면 결괴적으로 1개가 저장됩니다
우리가 원하는 결과물은 어떻게 되어야 하나요?
Thread 2개가 repo.save꺼지 왔다는것은 저장 요청이 총 2개라는것을 의미합니다
그렇다면 결과물로 id거 다른 2개의 객체가 map에 들어있어야 합니다

Reply:
아....제가 위에 적은 예시처럼 작성하면 데이터가 덮어씌워지는 이슈가 발생하지는 않더라도, 첫번째 요청만 처리하게 되겠군요. 그렇다면 동시에 들어온 요청이 sequence를 동시에 증가시켜 버리는 것부터가 문제같네요..! 애초에 sequence라는 것 자체가 thread-safe해야 할 것 같아서 AtomicLong으로 바꿔줘야 할 것 같습니다! 

맞을까요..두근